Rework volume reconstruction

jsafrane opened the pull request:
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Right now, volume reconstruction from pod directories is done *after* the desired state of world is populated, i.e. when kubelet is able to talk to the API server and long after kubelet starts running Pods, either discovered from static pod yamls or from the API server. At this point, kubelet does not know *what SELinux mount option* was used to mount the volumes, therefore it may provide volumes with wrong SELinux mounts to Pods.

Split volume reconstruction into multiple phases:

1. When kubelet starts, scan pod directories and reconstruct the volumes from there. The reconstructed volumes are marked as "uncertain" - kubelet must either call `MountDevice()` + `SetUp()` when they're needed or `TearDown()` or `UnmountDevice` if they're to be deleted. SELinux mount options reconstruction will be added in a subsequent PR.
3. (At this point allow starting new pods, e.g. from static pod yaml files)
4. When kubelet can reach the API server, finish reconstruction of volumes from node status.
5. When desired state of world is populated from the API server (and therefore kubelet knows what volume mounts are needed and what not), unmount (`TearDown` + `UnmountDevice`) the reconstructed volumes that are not needed.

#### Which issue(s) this PR fixes:
It's a prerequisite for https://github.com/kubernetes/enhancements/issues/1710. Kubelet must know (reconstruct) SELinux mount option used to mount the volume **before** starting Pods to be able to ensure that volumes are mounted with the SELinux label that Pods actually need.

I'm updating the KEP [here](https://github.com/kubernetes/enhancements/pull/3548).
 
#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
https://github.com/kubernetes/enhancements/pull/3548
<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1710-selinux-relabeling
```

jsafrane committed:
Rework volume reconstruction

Subsequent SELinux work (see http://kep.k8s.io/1710) will need
ActualStateOfWorld populated around the time kubelet starts mounting
volumes.

Therefore reconstruct volumes before starting reconciler, but do not depend
on the desired state of world populated nor node.status - both need a
working API server, which may not be available at that time.

All reconstructed volumes are marked as Uncertain and reconciler will sort
them out - call SetUp to ensure the volume is really mounted when a pod
needs the volume or call TearDown then there is no such pod.

Finish the reconstruction when the API server becomes available:
- Clean up volumes that failed reconstruction and are not needed.

- Update devicePath of reconstructed volumes from node.status. Make sure
  not to overwrite devicePath that may have been updated when the volume
  was mounted by reconcile().

Hiding all this rework behind SELinuxMountReadWriteOncePod FeatureGate,
just to make sure we have a way back if this commit is buggy.

jsafrane committed:
Remove SyncReconstructedVolume call

With the new reconstruction, AWS.MarkVolumeAsMounted will update outer spec
name with the correct value from Pod.

jsafrane committed:
Move all volume reconstruction code into separate files

There is no code change, just moving code around and preparing for the
subsequent commit.

jsafrane committed:
Refactor ConstructVolumeSpec

Return a struct from ConstructVolumeSpec to be able to add more fields to
it later.

jsafrane commented:
/test help
looking a disruptive test

jsafrane commented:
/test pull-kubernetes-e2e-gce-storage-disruptive

jsafrane commented:
/test pull-kubernetes-e2e-gce-storage-disruptive

jsafrane commented:
CSI disruptive tests passed

xing-yang commented:
/assign @jingxu97 @gnufied

jsafrane commented:
/test pull-kubernetes-e2e-capz-windows-containerd
/test pull-kubernetes-integration

The failures look like flakes, at least I don't see any volumes involved in TestDSCUpdatesPodLabelAfterDedupCurHistories and  azure conformance tests.

jsafrane commented:
/test pull-kubernetes-e2e-gce-csi-serial
@jingxu97 This should run disruptive tests with gcp-compute-persistent-disk-csi-driver, however, with disabled `SELinuxMountReadWriteOncePod` feature gate.

gnufied commented on pkg/kubelet/volumemanager/cache/actual_state_of_world.go:
```diff
@@ -501,6 +507,24 @@ func (asw *actualStateOfWorld) MarkDeviceAsUnmounted(
 
+func (asw *actualStateOfWorld) UpdateReconstructedDevicePath(volumeName v1.UniqueVolumeName, devicePath string) {
+	asw.RLock()
+	defer asw.RUnlock()
```
read lock for write operation?

jsafrane replied:
Fixed and squashed to "Rework volume reconstruction"

gnufied commented on pkg/kubelet/volumemanager/reconciler/reconstruct_new.go:
```diff
@@ -0,0 +1,241 @@
+		// so they can proceed with MountDevice / SetUp.
+		rc.desiredStateOfWorld.MarkVolumesReportedInUse(rc.volumesNeedReportedInUse)
+		rc.volumesNeedReportedInUse = nil
+	}
```
nit: since most of the code belongs to reconciler and not recontruction, we can have this function in `reconciler_new.go` or even in existing reconciler.go file.

jsafrane replied:
Added `reconciler_new.go` and `reconciler_common.go`, keeping the old code in `reconciler.go`, so we can rename `reconciler_new.go` to `reconciler.go` (overwriting the old code) when the feature is GA.

Added as a separate commit.

jingxu97 replied:
Just feel this mark volume in use logic is not implemented that straightforward. If possible, maybe add some explanation in Kep or somewhere?

gnufied commented on pkg/kubelet/volumemanager/reconciler/reconstruct_new.go:
```diff
@@ -0,0 +1,241 @@
+			if volumeID != attachedVolume.Name {
+				continue
+			}
+			rc.actualStateOfWorld.UpdateReconstructedDevicePath(volumeID, attachedVolume.DevicePath)
```
Is this racy with regular device path update that can happen with parallely running `mountOrAttachVolumes` function?

jsafrane replied:
That's why UpdateReconstructedDevicePath says "The ASW is updated only when the volume is still uncertain"

jsafrane commented on pkg/kubelet/volumemanager/reconciler/reconstruct_new.go:
```diff
@@ -0,0 +1,241 @@
+	}
+}
+
+func (rc *reconciler) readyToUnmount() bool {
```
Added a comment with similar meaning below. Squashed to "Rework volume reconstruction"

jsafrane committed:
Move new reconciler logic into reconciler_new.go

Move reconciler logic from reconstruct{new}.go to:
- reconciler.go - only the functionality used by the current (old)
  reconciler.
- reconciler_new.go - only the functionality used by the new reconciler.
- reconciler_common.go - common functions.

jsafrane committed:
Reshuffle functions between reconstruct and reconstruc_common

Move common functions to reconstruc_common.go and functions used only for
the current (old) reconstruction to reconstruct.go

jsafrane commented on pkg/kubelet/volumemanager/reconciler/reconstruct_new.go:
```diff
@@ -0,0 +1,241 @@
+
+	klog.V(2).InfoS("Orphan volume cleanup finished")
+	// Clean the cache, cleanup is one shot operation.
+	rc.volumesFailedReconstruction = make([]podVolume, 0)
```
I need to think about this. We probably don't want to retry in every reconcile loop (every 100ms), there should be some backoff.

jsafrane replied:
And perhaps we could try to reconstruct the volume again here few times...

gnufied replied:
Ack - we can fix this as a follow up https://github.com/kubernetes/kubernetes/issues/111933 (as long as we don't forget).

gnufied commented on pkg/kubelet/volumemanager/reconciler/reconstruct_new.go:
```diff
@@ -0,0 +1,197 @@
+	klog.V(2).InfoS("Volume reconstruction finished")
+}
+
+func (rc *reconciler) updateStatesNew(reconstructedVolumes map[v1.UniqueVolumeName]*globalVolumeInfo) {
```
We should have unit tests on the line of existing `func TestSyncStates(t *testing.T) {` when this feature gate is enabled. I know that behaviour will be subtly different, but we should still cover the failure cases.

jsafrane commented:
/test pull-kubernetes-e2e-gce-csi-serial

jsafrane committed:
Add unit tests

gnufied commented on pkg/kubelet/volumemanager/reconciler/reconstruct_new_test.go:
```diff
@@ -0,0 +1,261 @@
+			// Act
+			rcInstance.reconstructVolumes()
+			if tc.postSyncStatCallback != nil {
+				err := tc.postSyncStatCallback(rcInstance, fakePlugin)
```
These tests don't use this callback. May as well remove them?

jsafrane replied:
Fixed

gnufied commented on pkg/kubelet/volumemanager/reconciler/reconstruct_new_test.go:
```diff
@@ -0,0 +1,261 @@
+			expectedVolumesNeedDevicePath:       []string{},
+			expectedVolumesFailedReconstruction: []string{"pvc-abcdef"},
+		},
+	}
```
We are missing a test that was included in fix - https://github.com/kubernetes/kubernetes/pull/110670 . For this PR, I know the test will be somewhat simpler, because all we need to check is if volume is in DSOW, whether it makes it into ASOW.

jsafrane replied:
I'm not sure I understand, the new `reconstructVolumes` does not check DSOW at all, because it's not yet populated. Volume reconstruction adds all reconstructed volumes to ASOW and failed to `volumesFailedReconstruction`. `TestCleanOrphanVolumes` then checks a failed volume is / is not cleaned by `cleanOrphanVolumes` if it is not / is in DSOW.

gnufied replied:
I was looking for a test that triggers following scenarios:

1. Volume gets picked during reconsturction and added to ASOW (verify that)
2. Mount fails and verify that volume is still in uncertain state in ASOW.

jsafrane replied:
1. Volume gets picked during reconstuction and added to ASOW 
2. DSOW gets populated
3. reconciler tries to mount, mount fails
4. check that it's still uncertain in ASOW?

I can try to do that. But I don't see how it's better than two already existing tests:

* 1. Volume gets picked during reconstuction and added to ASOW 
  2. check that it's uncertain in ASOW


* 1. volume is uncertain in ASOW and it is in DSOW
  2. reconciler tries to mount and fails
  3. check that it's still uncertain in ASOW

jsafrane replied:
Now I see! Adding unit tests that checks `asw.foundDuringReconstruction` is set for reconstructed volumes.

jsafrane replied:
New unit tests added + I actually fixed regression of #110670. Good catch!

jsafrane committed:
Add unit test for failed mount after reconstruction

To preserve fix in https://github.com/kubernetes/kubernetes/pull/110670,
add an unit test that check a volume is *uncertain* even after final mount
error when it was reconstructed.

And actually fix a regression introduced in the previous patch.

gnufied reviewed (commented):
/lgtm

gnufied commented:
Just putting on hold in case @jingxu97 or @msau42 wants to take a stab at it.

/hold

jsafrane commented:
/test pull-kubernetes-unit
A/D controller flake
/test pull-kubernetes-integration
kubectl timeouts

jingxu97 commented on pkg/kubelet/volumemanager/cache/actual_state_of_world.go:
```diff
@@ -636,7 +660,16 @@ func (asw *actualStateOfWorld) AddPodToVolume(markVolumeOpts operationexecutor.M
 
 	podObj, podExists := volumeObj.mountedPods[podName]
-	if !podExists {
+
```
not your change, but why we use podNmae as key here? Why not podUID? what if two different pods with the same name (old and new)?

jingxu97 replied:
actually I think we use Pod uid as pod unique name

jsafrane replied:
Yes, podName is UID. I'm not sure why it's so confusing.

jingxu97 commented on pkg/kubelet/volumemanager/reconciler/reconstruct_new.go:
```diff
@@ -0,0 +1,243 @@
+			rc.volumesFailedReconstruction = append(rc.volumesFailedReconstruction, volume)
+			continue
+		}
+		klog.V(4).InfoS("Adding reconstructed volume to actual state and node status", "podName", volume.podName, "volumeSpecName", volume.volumeSpecName)
```
use v(2)?

jsafrane replied:
This is actually copied from the old reconstruction, where it is at `V(4)`:
https://github.com/kubernetes/kubernetes/blob/3a79466ddde88e5161f1ca9754bebc62ef4ac8b9/pkg/kubelet/volumemanager/reconciler/reconstruct.go#L154.

Looking at logs, there is just V(2): `reconstruct_new.go:103] "Volume reconstruction finished"`. Since mount / unmount `MountVolume.TearDown succeeded for volume` is at `V(2)`, it makes sense to add reconstructed volumes there too. Added as a separate commit in both old/new reconstruction.

jingxu97 commented on pkg/kubelet/volumemanager/reconciler/reconstruct_new.go:
```diff
@@ -0,0 +1,243 @@
+	for _, gvl := range reconstructedVolumes {
+		err := rc.actualStateOfWorld.MarkVolumeAsAttached(
+			//TODO: the devicePath might not be correct for some volume plugins: see issue #54108
+			gvl.volumeName, gvl.volumeSpec, "" /* nodeName */, gvl.devicePath)
```
just to confirm nodeName is empty?

jsafrane replied:
This is copied from the old reconstruction: https://github.com/kubernetes/kubernetes/blob/3a79466ddde88e5161f1ca9754bebc62ef4ac8b9/pkg/kubelet/volumemanager/reconciler/reconstruct.go#L142-L143

`nodeName` in `MarkVolumeAsAttached` is then not used:
https://github.com/kubernetes/kubernetes/blob/e0f3e5c4575f2ad0c1a62d2b24c5ebe8483da5c0/pkg/kubelet/volumemanager/cache/actual_state_of_world.go#L361-L364

The parameter is there to satisfy `ActualStateOfWorldAttacherUpdater` [interface](https://github.com/kubernetes/kubernetes/blob/fc245b339b3a72307435f51c02e783eac7b66825/pkg/volume/util/operationexecutor/operation_executor.go#L240).

Now that I am looking at it, adding real node name is simple and would make the code more robust. Fixing in a new commit in both old/new reconstruction.

jingxu97 commented on pkg/kubelet/volumemanager/reconciler/reconstruct_new.go:
```diff
@@ -0,0 +1,243 @@
+			}
+			klog.V(4).InfoS("Volume is marked as uncertain and added into the actual state", "pod", klog.KObj(volume.pod), "podName", volume.podName, "volumeName", volume.volumeName)
+		}
+		// If the volume has device to mount, we mark its device as mounted.
```
mark its device as uncertain?

jsafrane replied:
Fixed

jingxu97 commented on pkg/kubelet/volumemanager/reconciler/reconstruct_new.go:
```diff
@@ -0,0 +1,243 @@
+				klog.ErrorS(err, "Could not mark device is uncertain to actual state of world", "volumeName", gvl.volumeName, "deviceMountPath", deviceMountPath)
+				continue
+			}
+			klog.V(4).InfoS("Volume is marked device as uncertain and added into the actual state", "volumeName", gvl.volumeName, "deviceMountPath", deviceMountPath)
```
use V(2)?

jsafrane replied:
Fixed

jingxu97 commented:
overall look good to me. Just a few small questions.

jsafrane committed:
Add node name parameter

Add nodeName to MarkVolumeAsAttached. MarkVolumeAsAttached implementation
in kubelet does not use the parameter, but it could do that in the future.

jsafrane committed:
Increase verbosity of volume reconstruction messages

Add volume reconstruction logs to V(2) to see initial kubelet
ActualStateOfWorld after kubelet start. Kubelet logs SetUp / TearDown
events at V(2) already, so we can track the whole volume mount state in
V(2) logs.

jsafrane commented:
@jingxu97 Addresses your comments, PTAL.

gnufied commented:
/hold cancel

gnufied commented:
/retest

gnufied commented:
/lgtm
